jsc#SLE-23788, Enhance WMP for a smooth handover to SAP's systemd

nick-wang opened the pull request:
* check.sh: Add unprotect_list check of subcgroups
  * check.sh: Fix a wmp check bug of memory_low_children
  * Skip systemd managed processes for jsc#PM-3309

nick-wang committed:
Skip systemd managed processes jsc#PM-3309

nick-wang committed:
Add switch f and avoid empty DBus message error.

nick-wang committed:
check.sh: Fix a wmp check bug of memory_low_children

nick-wang committed:
check.sh: Add unprotect_list check of subcgroups

scmschmidt reviewed (approved):
Looks good to me, but I'm not a good C programmer!

scmschmidt reviewed (commented):
Looks good

scmschmidt reviewed (commented):
Looks good.

nick-wang commented:
> Looks good to me, but I'm not a good C programmer!

@scmschmidt Thanks for the review! You are quite humble~

Werkov commented on src/main.c:
```diff
@@ -39,6 +39,41 @@ static inline void freep(void *p) {
 	free(*(void **)p);
 }
 
+int _already_in_systemd(pid_t pid, const char *target_slice) {
```
nit: Maybe `already_in_slice`?

nick-wang replied:
Changed to already_in_slice in the new commit.

Werkov commented on src/main.c:
```diff
@@ -39,6 +39,41 @@ static inline void freep(void *p) {
+	if (r < 0)
+		return r;
+
+	r = snprintf(pattern, UNIT_NAME_LEN + 2, "/%s/", target_slice);
```
I think absolute pattern would be better in this case, such as `"0::/%s/"`.
IOW the check should match where the processes would end up after migration by the capture program.
(Or is the flexibility intended? E.g. on `/foo.slice/TARGER.slice` or `/foo.slice/bar.service/TARGET.slice/baz`?)

nick-wang replied:
In fact, I keep the flexibility on purpose.
`sapwmp-capture` is manually added in SAP start script. So there is no risk the `/foo.slice/bar.service/TARGET.slice/baz` being affected accidentally.
If a user intend to add the capture program and use `TARGET.slice` as a sub-slice in any reason, i think we should skip capture process in this case.

BTW, both methods are/should work on the current SAP use scenarios.

Werkov replied:
I forgot in my previous comment that systemd requires slice names to encode the hierarchy, i.e. the first example can't exists (would be `/foo.slice/foo-TARGET.slice`), so the relative match is absolute in this case.
The second example represents a case when cgroups are further delegated to a service and it creates own subcgroup with the matching name (`TARGET.slice`). This does not happen in our use case and even malicious service can't gain anything by masking as such.

So, I think the current match is good in the end.

nick-wang replied:
Thanks for the information! I don't know that before, it is good learning.

Werkov commented on src/main.c:
```diff
@@ -39,6 +39,41 @@ static inline void freep(void *p) {
 
+int _already_in_systemd(pid_t pid, const char *target_slice) {
+	char path[PATH_MAX];
+	char buf[LINE_LEN];
```
Since the cgroup file lists paths, I'd use `PATH_MAX` for this buffer and keep `LINE_LEN` for configs only.
+less likely some long cgroup paths would be parsed broken.

nick-wang replied:
OK. Acutally PATH_MAX(4096 on my SLE) is longer than LINE_LEN(1024), it is safe to use.
Changed to PATH_MAX.

Werkov commented on src/main.c:
```diff
@@ -39,6 +39,41 @@ static inline void freep(void *p) {
+
+	f = fopen(path, "r");
+	if (!f)
+		return -errno;
```
Is this intentional? The function caller expects a boolean value.

nick-wang replied:
Keep the `-errno` and change to return value conditions check in caller, just like `read_stat()`

Werkov commented on src/main.c:
```diff
@@ -39,6 +39,41 @@ static inline void freep(void *p) {
+			goto final;
+		}
+	}
+	r = 0;
```
nit: Move before the loop and replace `goto` with `break`?

nick-wang replied:
Changed in the new commit.

Werkov commented on src/main.c:
```diff
@@ -202,7 +237,10 @@ int collect_pids(pid_t **rpids) {
 		for (char **p = config.parent_commands.list; *p; p++) {
 			if(!strcmp(comm, *p)) {
-				pids[n_pids++] = pid;
+				if (!force && _already_in_systemd(pid, config.slice))
```
Given possible failure in `_already_in_systemd`, I'd propagate it through whole `collect_pids` (similar to `read_stat`) failure to be on the safe side when such an unexpected error happens.

nick-wang replied:
Changed in the new commit.

Werkov commented on src/main.c:
```diff
@@ -237,25 +275,26 @@ static int make_scope_name(char *buf) {
 	fprintf(stderr, "\n");
 	fprintf(stderr, "       -h	Show this help\n");
 	fprintf(stderr, "       -v	Verbose logging (for debugging)\n");
+	fprintf(stderr, "       -f	Force protect all valid processes(including systemd launched)\n");
```
nit: s/protect/migrate/ or s/protect/capture/

nick-wang replied:
Changed to capture in the new commit.

Werkov commented on src/main.c:
```diff
@@ -285,11 +327,11 @@ int main(int argc, char *argv[]) {
 		exit_error(EXIT_FAILURE, n_pids, "Failed collecting PIDs");
 	else if (n_pids == 0)
-		exit_error(EXIT_SUCCESS, 0, "Empty capture");
+		return EXIT_SUCCESS;
```
nit: I'd keep the `"Empty capture"` message here for easier troubleshooting (despite it's implied by other messages).

nick-wang replied:
Changed in the new commit.

Werkov commented on scripts/wmp-check.sh:
```diff
@@ -281,23 +282,26 @@ function get_SAP_slice_data() {
+        if [ ${error} -ne 0 ] ; then
             SAP_slice_data['memory_low_children']='false'
+        else
+            SAP_slice_data['memory_low_children']='ok'
```
Good catch!

Werkov reviewed (changes requested):
wmp-check.sh changes look good to me.
I have two main remarks for capture program: a) absolute cgroup path matching, b) error handling (details + some nits are in inlined).

nick-wang committed:
Polish pull request based on review comments.

1. Move LINE_LEN back and use PATH_MAX
2. Rename _already_in_systemd to already_in_slice
3. Use break instead of goto in already_in_slice
4. Output "Empty capture" when no process captured

nick-wang commented:
@Werkov Thanks for the review. I fixed all comments but one(reason as above comment) in the new commit.

Werkov reviewed (commented):
Thanks for the fixups. Please check my reasoning regarding the relative vs absolute matching. If you agree, we're good to go.

nick-wang commented:
> Thanks for the fixups. Please check my reasoning regarding the relative vs absolute matching. If you agree, we're good to go.

Thank you. Always can learn something from you.
I didn't find similar case like `foo-TARGET.slice` in my experience. It is good to know, i will make an experiment tomorrow.

So if we have positive feedback after testing, we can merge this.

Werkov reviewed (approved):
LGTM and I agree with proposed steps.

nick-wang commented:
> LGTM and I agree with proposed steps.

Thanks for the review and comments!

liangxin1300 commented:
@nick-wang 
Thanks!

nick-wang committed:
wmp-check: raise error when user not configure MemoryLow of target slice in digital.

nick-wang commented:
With the test of @scmschmidt , found if customer  (incorrectly!!!!) set MemoryLow of SAP.slice(TARGET.slice) left with infinity. Error message will show:
```
/usr/sbin/wmp-check: Zeile 701: [: infinity: Ganzzahliger Ausdruck erwartet.
[FAIL]  MemoryLow (infinity) is smaller then the current allocated memory (3143348224) for SAP.slice!
        -> Check if this is an expected situation.
```
In this case, user is not set MemoryLow of SAP.slice(TARGET.slice)

nick-wang commented:
> With the test of @scmschmidt , found if customer (incorrectly!!!!) set MemoryLow of SAP.slice(TARGET.slice) left with infinity. Error message will show:
> 
> ```
> /usr/sbin/wmp-check: Zeile 701: [: infinity: Ganzzahliger Ausdruck erwartet.
> [FAIL]  MemoryLow (infinity) is smaller then the current allocated memory (3143348224) for SAP.slice!
>         -> Check if this is an expected situation.
> ```

So i add a commit to raise the issue of incorrectly MemoryLow configuration. @scmschmidt  How do you think of this?

scmschmidt reviewed (commented):
line 720: print_fail "MemoryLow of ${SAP_slice_data['name']} (${SAP_slice_data['MemoryLow']}) is not digital!" "Configure MemoryLow to a number."

"...is not a nnumeric value" would be a better phrase then "is not digital".

nick-wang committed:
wmp-check: Polish the phrase of error output.

nick-wang commented:
I changed the phrase in the last commit.

Considering the different timezone, it is likely my last pr here.
I think i can't access SUSE github repo anymore.

Thanks for the review. It was my pleasure and honor to work with you @Werkov and @scmschmidt
